### PR TITLE
fix(ffi): avoid false close error and stranded lease on panic

### PR DESCRIPTION
**Pull request by powerslider**
## Why this should be merged

Should be merged in #1976 or parts of it that resolve correctness issues and/or introduce better coverage for regressions.

## How this works

Resolve correctness issues found while reviewing the force-close work:

- `Database.Close` no longer returns `ErrActiveKeepAliveHandles` when force-closing with no live handles under an already-cancelled context. `waitForKeepAlives` is now the sole check for remaining handles and returns nil once the count drains, regardless of ctx state.
- `lease.release` releases the lease even when its callback panics, so a panicking free can no longer strand the lease and block `Close` forever.
- `iteratorHandle.Drop` marks the handle dropped before freeing, matching `handle[T].Drop`, so a panic in free cannot leave a retry able to double-free.
- `removeAndDecr` defers its unlock, matching `attachUnregistered` and preventing a stranded registry lock on panic.
- Restore `fmt.Errorf` for `errFreeingValue` wraps where the inner error is the cause rather than an independent error.
- Document the `Iterator` single-reader contract and that force-close does not auto-drop verified `RangeProof`s.

## How this was tested

Add tests for the force-close edge cases, panic-safe release, closed-registry attach, `Iterator`/`Reconstructed` finalizer back-stops, a live `RangeProof` under force-close, and partial-drain-then-retry.

## Breaking Changes
none

Signed-off-by: Tsvetan Dimitrov (tsvetan.dimitrov@avalabs.org)

### CODE REVIEW

File: ffi/firewood.go [hunk 1]
```diff
@@ -538,16 +539,11 @@ func (db *Database) Close(ctx context.Context, opts ...CloseOption) error {
 	var forcedDropErr error
 	if cfg.forceCloseHandles {
 		forcedDropErr = db.keepAlives.closeAndForceDrop(ctx)
-		if ctx.Err() != nil {
```
Comment (powerslider) [inline ffi/firewood.go:541]: Force-close under a cancelled context returns `ErrActiveKeepAliveHandles` even when no handles are outstanding. An extra `ctx.Err()` check fires before `waitForKeepAlives`, which already returns nil when nothing is left to wait for. That is why I removed the check and let `waitForKeepAlives` decide.

File: ffi/keepalive.go [hunk 2]
```diff
@@ -339,11 +340,19 @@ func (l *lease) attachUnregistered(registry *keepAliveRegistry) {
...
-	err := attemptDisown()
+	// Release on panic too, so a panicking callback cannot strand the
+	// lease and block Close forever.
+	defer func() {
+		if p := recover(); p != nil {
+			l.releaseLocked()
+			panic(p)
+		}
+	}()
+	err = attemptDisown()
```
Comment (powerslider) [inline ffi/keepalive.go:355]: If `attemptDisown` panics, `releaseLocked` never runs, so the lease is left registered with its count still held, and any later graceful close waits on it forever. The fix is to release the lease on the panic path as well and then re-raise. That follows the reasoning the PR already applies elsewhere: a Close that hangs is worse than a free that errored. Let me know if this line of thought makes sense.

**Review by rkuris — APPROVED** [review]
Reviewed all six fixes — the concurrency reasoning checks out (waitForKeepAlives as sole drain authority with the dual-select, panic-safe release defer ordering, dropped-before-free matching handle[T].Drop, deferred unlock in removeAndDecr, and the %w: %w restoration). Lint and `go test -race` are clean locally. Thanks for catching these!

Two non-blocking notes, fine as follow-ups:
- `TestCloseForceDoesNotDropRangeProof` waits out the full 1s context on every run since the RangeProof is never auto-dropped; a ~1ms timeout ctx would assert the same failure without the wall-clock cost.
- A one-line comment on why that test uses `newVerifiedRangeProof` + `db.VerifyRangeProof` (to put a lease on the proof) would help readers — especially since calling `VerifyRangeProof` twice on the same proof panics in `attachUnregistered`.
